src/mbr.c

```diff
@@ -0,0 +1,458 @@
+
+	get_chs(&raw_entry->chs_start, raw_entry->partition_start, &geometry);
+
+	get_chs(&raw_entry->chs_end,
+			raw_entry->partition_start + raw_entry->partition_size - 1,
+			&geometry);
+out:
+	return res;
+}
+
+gboolean r_mbr_switch_get_inactive_partition(const gchar *device,
+		struct mbr_switch_partition *partition,
+		const gchar *region_start, const gchar *region_size,
+		GError **error)
+{
+	gboolean res = FALSE;
+	struct mbr mbr;
+	GError *ierror = NULL;
+	struct partition_tbl_entry *boot_part;
+	struct mbr_switch_partition region;
+	guint sector_size;
+	gint fd;
+
+	fd = g_open(device, O_RDONLY);
+
+	sector_size = get_sectorsize(fd);
+
+	res = get_region(fd, &region, region_start, region_size, sector_size,
+			&ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	res = read_mbr(fd, &mbr, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror,
+				"failed to read mbr:");
+		goto out;
+	}
+	res = FALSE;
+
+	/* check if region overlaps with any partition */
+	if (!is_region_free(&region, mbr.partition_table, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	boot_part = &mbr.partition_table[BOOT_PARTITION_ENTRY];
+
+	if (boot_part->partition_start == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out;
+	}
+
+	if ((guint32)(region.start / sector_size) ==
+	    boot_part->partition_start) {
+		partition->start = region.start + region.size / 2;
+	}
+	else if ((guint32)((region.start + region.size / 2) / sector_size) ==
+	         boot_part->partition_start) {
+		partition->start = region.start;
+	}
+	else {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"boot partition's start address does match region "
+				"configuration");
+		goto out;
+	}
+	partition->size = region.size / 2;
+
+	res = TRUE;
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+gboolean r_mbr_switch_clear_partition(const gchar *device,
+		const struct mbr_switch_partition *dest_partition,
+		GError **error)
+{
+	gboolean res = FALSE;
+	static gchar zerobuf[512] = {};
+	gint clear_size = sizeof(zerobuf);
+	guint clear_count = 0;
+	gint tmp_count = 0;
+	gint fd;
+
+	fd = g_open(device, O_RDWR);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening device failed: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
+	if (lseek(fd, dest_partition->start, SEEK_SET) !=
+	    (off_t)dest_partition->start) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to set file to position %lu: %s",
+				dest_partition->start, g_strerror(errno));
+		goto out;
+	}
+
+	while (clear_count < dest_partition->size) {
+		if ((dest_partition->size - clear_count) < sizeof(zerobuf))
+			clear_size = (dest_partition->size - clear_count);
+
+		tmp_count = write(fd, zerobuf, clear_size);
+
+		if (tmp_count != clear_size)
+			break;
```

Comment:
This could return EINTR, or less than the requested bytes. In that case, you still need to count the written bytes.